crypto/x509/v3_utl.c

```diff
@@ -978,19 +978,27 @@ int X509_check_ip_asc(X509 *x, const char *ipasc, unsigned int flags)
 
 char *ipaddr_to_asc(unsigned char *p, int len)
 {
+    /*
+     * 40 is enough space for the longest IPv6 address + nul terminator byte
+     * XXXX:XXXX:XXXX:XXXX:XXXX:XXXX:XXXX:XXXX\0
+     */
     char buf[40], *out;
+    int i = 0, remain = 0, bytes = 0;
 
     switch (len) {
     case 4: /* IPv4 */
         BIO_snprintf(buf, sizeof(buf), "%d.%d.%d.%d", p[0], p[1], p[2], p[3]);
         break;
         /* TODO possibly combine with static i2r_address() in v3_addr.c */
     case 16: /* IPv6 */
-        for (out = buf; out < buf + 8 * 3; out += 3) {
-            BIO_snprintf(out, 3 + 1, "%X:", p[0] << 8 | p[1]);
+        for (out = buf, i = 8, remain = sizeof(buf);
+             i-- > 0 && bytes >= 0;
+             remain -= bytes, out += bytes) {
+            const char *template = (i > 0 ? "%X:" : "%X");
+
+            bytes = BIO_snprintf(out, remain, template, p[0] << 8 | p[1]);
             p += 2;
         }
-        out[-1] = '\0';
```

Comment:
I suggest keeping an explicit NUL termination just in case `BIO_snprintf()` failed (indicated by `bytes < 0`)
because in this case I fear that it is not guaranteed that the string is properly terminated:
`buf[out - bytes] = '\0';`

In case the `BIO_printf()` output was truncated according to the buffer limit it would return `remain`, which would terminate the loop (and thus the output string) prematurely, 
but this cannot happen because the buffer is surely large enough.

Reply:
Do you see the comment above the buffer?  In IPv6 address will be a maximum of 39 characters (bytes in this case) when written out, so there's absolutely enough space for the terminating NUL byte.

Reply:
Yes, but as I wrote (according to the man page of `snprintf` etc., so likely also holds for `BIO_snprintf`):
On potential truncation (not possible here), the function will return `reminder`,
and on an output error it returns a negative value.
And unfortunately the documentation is not really clear (at least to me) whether the string is properly terminated in such (non-success) cases.
Yet we may be safe here because for string output no output error can occur.